File: apps/ios/FishKit/Sources/ChatCore/Logic/ChatSelectors.swift
import Foundation

public enum ChatSelectors {
    public static func compareChatMessages(
        _ left: ChatMessageState,
        _ right: ChatMessageState
    ) -> Bool {
        let leftDate = timestamp(left.createdAt)
        let rightDate = timestamp(right.createdAt)
        if leftDate != rightDate { return leftDate < rightDate }
        return left.id.compare(right.id) == .orderedAscending
    }

    public static func mergeChatMessage(
        _ current: [ChatMessageState],
        _ incoming: ChatMessageState,
        localRequestId: String? = nil
    ) -> [ChatMessageState] {
        let requestId = localRequestId ?? incoming.clientRequestId
        guard let index = current.firstIndex(where: {
            $0.id == incoming.id
                || $0.clientRequestId == incoming.clientRequestId
                || $0.clientRequestId == requestId
        }) else {
            return (current + [incoming]).sorted(by: compareChatMessages)
        }

        let existing = current[index]
        let incomingAttachments = incoming.attachments ?? incoming.images
        let existingAttachments = existing.attachments ?? existing.images
        let attachments = incomingAttachments?.isEmpty == false
            ? incomingAttachments
            : existingAttachments
        var merged = incoming
        merged.senderDisplayName = incoming.senderDisplayName ?? existing.senderDisplayName
        merged.attachments = attachments
        merged.images = attachments
        merged.gif = incoming.gif ?? existing.gif
        merged.stickerId = incoming.stickerId ?? existing.stickerId
        guard merged != existing else { return current }

        var next = current
        next[index] = merged
        return next.sorted(by: compareChatMessages)
    }

    public static func mergeReadState(
        _ current: [ChatReadState],
        _ incoming: ChatReadState
    ) -> [ChatReadState] {
        guard let index = current.firstIndex(where: { $0.userId == incoming.userId }) else {
            return current + [incoming]
        }
        guard current[index] != incoming else { return current }
        var next = current
        next[index] = incoming
        return next
    }

    public static func getOutgoingMessageStatus(
        _ message: ChatMessageState,
        messages: [ChatMessageState],
        participantReadState: ChatReadState?
    ) -> OutgoingMessageStatus {
        if isAtOrAfterMessage(
            participantReadState?.lastReadMessageId,
            targetMessageId: message.id,
            messages: messages
        ) {
            return .read
        }
        if isAtOrAfterMessage(
            participantReadState?.lastDeliveredMessageId,
            targetMessageId: message.id,
            messages: messages
        ) {
            return .delivered
        }
        return .sent
    }

    public static func countUnreadMessages(
        _ messages: [ChatMessageState],
        currentUserId: String,
        currentUserReadState: ChatReadState?
    ) -> Int {
        getUnreadMessageSummary(
            messages,
            currentUserId: currentUserId,
            currentUserReadState: currentUserReadState
        ).count
    }

    public static func getUnreadMessageSummary(
        _ messages: [ChatMessageState],
        currentUserId: String,
        currentUserReadState: ChatReadState?
    ) -> UnreadMessageSummary {
        let lastReadIndex = currentUserReadState?.lastReadMessageId.flatMap { marker in
            messages.firstIndex(where: { $0.id == marker })
        } ?? -1
        let unread = messages.enumerated().compactMap { index, message in
            index > lastReadIndex
                && message.senderId != currentUserId
                && message.deletedAt == nil
                ? message
                : nil
        }
        return UnreadMessageSummary(
            count: unread.count,
            oldestUnreadAt: unread.first?.createdAt,
            latestUnreadMessageId: unread.last?.id
        )
    }

    public static func getMessageSnippet(_ message: ChatMessageState) -> String {
        if message.deletedAt != nil { return "Message deleted" }
        let body = message.body.trimmingCharacters(in: .whitespacesAndNewlines)
        if body.isEmpty, message.stickerId != nil { return "Sticker" }
        if body.isEmpty, message.gif != nil { return "GIF" }
        let attachments = message.attachments ?? message.images ?? []
        if body.isEmpty, !attachments.isEmpty {
            if attachments.count == 1 {
                return attachments[0].kind == "file" ? "File" : "Image"
            }
            return attachments.allSatisfy { $0.kind != "file" }
                ? "\(attachments.count) images"
                : "\(attachments.count) files"
        }
        guard body.count > 96 else { return body }
        return String(body.prefix(95)) + "…"
    }

    public static func toReplyPreview(
        _ message: ChatMessageState,
        currentUserId: String,
        participantName: String,
        currentUserName: String
    ) -> ReplyPreview {
        ReplyPreview(
            id: message.id,
            authorName: message.senderId == currentUserId ? currentUserName : participantName,
            snippet: getMessageSnippet(message)
        )
    }

    private static func isAtOrAfterMessage(
        _ markerMessageId: String?,
        targetMessageId: String,
        messages: [ChatMessageState]
    ) -> Bool {
        guard
            let markerMessageId,
            let targetIndex = messages.firstIndex(where: { $0.id == targetMessageId }),
            let markerIndex = messages.firstIndex(where: { $0.id == markerMessageId })
        else { return false }
        return markerIndex >= targetIndex
    }

    private static func timestamp(_ value: String) -> Date {
        if let date = try? Date(value, strategy: .iso8601) { return date }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.date(from: value) ?? .distantPast
    }
}

public func compareChatMessages(_ left: ChatMessageState, _ right: ChatMessageState) -> Bool {
    ChatSelectors.compareChatMessages(left, right)
}

public func mergeChatMessage(
    _ current: [ChatMessageState],
    _ incoming: ChatMessageState,
    localRequestId: String? = nil
) -> [ChatMessageState] {
    ChatSelectors.mergeChatMessage(current, incoming, localRequestId: localRequestId)
}

public func mergeReadState(_ current: [ChatReadState], _ incoming: ChatReadState) -> [ChatReadState] {
    ChatSelectors.mergeReadState(current, incoming)
}

public func getOutgoingMessageStatus(
    _ message: ChatMessageState,
    messages: [ChatMessageState],
    participantReadState: ChatReadState?
) -> OutgoingMessageStatus {
    ChatSelectors.getOutgoingMessageStatus(
        message,
        messages: messages,
        participantReadState: participantReadState
    )
}

public func countUnreadMessages(
    _ messages: [ChatMessageState],
    currentUserId: String,
    currentUserReadState: ChatReadState?
) -> Int {
    ChatSelectors.countUnreadMessages(
        messages,
        currentUserId: currentUserId,
        currentUserReadState: currentUserReadState
    )
}

public func getUnreadMessageSummary(
    _ messages: [ChatMessageState],
    currentUserId: String,
    currentUserReadState: ChatReadState?
) -> UnreadMessageSummary {
    ChatSelectors.getUnreadMessageSummary(
        messages,
        currentUserId: currentUserId,
        currentUserReadState: currentUserReadState
    )
}

public func getMessageSnippet(_ message: ChatMessageState) -> String {
    ChatSelectors.getMessageSnippet(message)
}

public func toReplyPreview(
    _ message: ChatMessageState,
    currentUserId: String,
    participantName: String,
    currentUserName: String
) -> ReplyPreview {
    ChatSelectors.toReplyPreview(
        message,
        currentUserId: currentUserId,
        participantName: participantName,
        currentUserName: currentUserName
    )
}

File: apps/android/feature/chat/src/main/kotlin/space/fishhub/android/feature/chat/state/ChatSelectors.kt
package space.fishhub.android.feature.chat.state

import space.fishhub.android.data.chat.model.ChatMessage
import space.fishhub.android.data.chat.model.ChatReadState
import space.fishhub.android.data.chat.model.ChatAttachmentKind
import java.time.Instant

val ChatMessageComparator: Comparator<ChatMessage> = Comparator { left, right ->
    val byTime = runCatching {
        Instant.parse(left.createdAt).compareTo(Instant.parse(right.createdAt))
    }.getOrElse {
        left.createdAt.compareTo(right.createdAt)
    }
    if (byTime == 0) left.id.compareTo(right.id) else byTime
}

fun mergeChatMessage(
    current: List<ChatMessage>,
    incoming: ChatMessage,
    localRequestId: String = incoming.clientRequestId,
): List<ChatMessage> {
    val existingIndex = current.indexOfFirst { message ->
        message.id == incoming.id ||
            message.clientRequestId == incoming.clientRequestId ||
            message.clientRequestId == localRequestId
    }
    if (existingIndex == -1) return (current + incoming).sortedWith(ChatMessageComparator)

    val existing = current[existingIndex]
    val merged = incoming.copy(
        senderDisplayName = incoming.senderDisplayName ?: existing.senderDisplayName,
        attachments = incoming.attachments.ifEmpty { existing.attachments },
        gif = incoming.gif ?: existing.gif,
        gifUnavailable = when {
            incoming.gif != null -> false
            incoming.gifUnavailable -> true
            else -> existing.gifUnavailable
        },
        stickerId = incoming.stickerId ?: existing.stickerId,
    )
    if (existing == merged) return current

    return current.toMutableList().apply { this[existingIndex] = merged }
        .sortedWith(ChatMessageComparator)
}

fun mergeReadState(
    current: List<ChatReadState>,
    incoming: ChatReadState,
): List<ChatReadState> {
    val existingIndex = current.indexOfFirst { it.userId == incoming.userId }
    if (existingIndex == -1) return current + incoming
    val existing = current[existingIndex]
    if (isEarlierTimestamp(incoming.deliveredAt, existing.deliveredAt) ||
        isEarlierTimestamp(incoming.readAt, existing.readAt) ||
        existing == incoming
    ) {
        return current
    }
    return current.toMutableList().apply { this[existingIndex] = incoming }
}

private fun isEarlierTimestamp(incoming: String?, current: String?): Boolean {
    if (current == null) return false
    if (incoming == null) return true
    return runCatching { Instant.parse(incoming).isBefore(Instant.parse(current)) }
        .getOrElse { incoming < current }
}

fun outgoingMessageStatus(
    message: ChatMessage,
    messages: List<ChatMessage>,
    participantReadState: ChatReadState?,
): OutgoingMessageStatus = when {
    isAtOrAfterMessage(participantReadState?.lastReadMessageId, message.id, messages) ->
        OutgoingMessageStatus.Read
    isAtOrAfterMessage(participantReadState?.lastDeliveredMessageId, message.id, messages) ->
        OutgoingMessageStatus.Delivered
    else -> OutgoingMessageStatus.Sent
}

private fun isAtOrAfterMessage(
    markerMessageId: String?,
    targetMessageId: String,
    messages: List<ChatMessage>,
): Boolean {
    if (markerMessageId == null) return false
    val targetIndex = messages.indexOfFirst { it.id == targetMessageId }
    val markerIndex = messages.indexOfFirst { it.id == markerMessageId }
    return markerIndex >= 0 && targetIndex >= 0 && markerIndex >= targetIndex
}

fun unreadMessageSummary(
    messages: List<ChatMessage>,
    currentUserId: String,
    currentUserReadState: ChatReadState?,
): UnreadMessageSummary {
    val lastReadIndex = currentUserReadState?.lastReadMessageId?.let { marker ->
        messages.indexOfFirst { it.id == marker }
    } ?: -1
    val unread = messages.filterIndexed { index, message ->
        index > lastReadIndex && message.senderId != currentUserId && message.deletedAt == null
    }
    return UnreadMessageSummary(
        count = unread.size,
        oldestUnreadAt = unread.firstOrNull()?.createdAt,
        latestUnreadMessageId = unread.lastOrNull()?.id,
    )
}

fun countUnreadMessages(
    messages: List<ChatMessage>,
    currentUserId: String,
    currentUserReadState: ChatReadState?,
): Int = unreadMessageSummary(messages, currentUserId, currentUserReadState).count

private const val MaxSnippetCodePoints = 96

fun messageSnippet(message: ChatMessage): String {
    if (message.deletedAt != null) return "Message deleted"
    val body = message.body.trim()
    if (body.isEmpty() && message.stickerId != null) return "Sticker"
    if (body.isEmpty() && message.gif != null) return "GIF"
    if (body.isEmpty() && message.gifUnavailable) return "GIF"
    if (body.isEmpty() && message.attachments.isNotEmpty()) {
        if (message.attachments.size == 1) {
            val attachment = message.attachments.first()
            return when (attachment.kind) {
                ChatAttachmentKind.Image -> "Photo"
                ChatAttachmentKind.File -> if (attachment.mimeType == "audio/mp4") {
                    "Voice message"
                } else {
                    "File"
                }
                ChatAttachmentKind.Unavailable -> "Attachment"
            }
        }
        return if (message.attachments.all { it.kind == ChatAttachmentKind.Image }) {
            "${message.attachments.size} photos"
        } else {
            "${message.attachments.size} attachments"
        }
    }
    val codePoints = body.codePoints().toArray()
    if (codePoints.size <= MaxSnippetCodePoints) return body
    return buildString {
        codePoints.take(MaxSnippetCodePoints - 1).forEach { appendCodePoint(it) }
        append('…')
    }
}

fun replyPreview(
    message: ChatMessage,
    currentUserId: String,
    participantName: String,
    currentUserName: String,
): ReplyPreview = ReplyPreview(
    id = message.id,
    authorName = if (message.senderId == currentUserId) currentUserName else participantName,
    snippet = messageSnippet(message),
)
